### ISSUE TITLE
IDE functionality is broken by stack overflow in directives when a file is full of comments

### ISSUE THREAD

**Issue by kubukoz** [VirtusLab/scala-cli#1603]
**Version(s)**

0.1.17

**Describe the bug**

The BSP server logs are full of stack traces. This seems to affect IDE functionality by showing stale diagnostics:

<img width="308" alt="image" src="https://user-images.githubusercontent.com/894884/203627250-3bd77a62-6a39-48b4-a9f3-95577ddfe25c.png">

which is fixed the moment you delete a problematic file:

<img width="941" alt="image" src="https://user-images.githubusercontent.com/894884/203627285-9cbad2b5-8c1a-4ac5-ac79-a514e283d43a.png">

**To Reproduce**

Open a project in Metals which contains a file with over 100 lines of comments. For example:

<details>
<summary>lotsa comments inside</summary>

```scala
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
// aaaaaaaaaaaaa
```

</details>

It may not sound realistic, but sometimes you comment out an entire file and that does surface the error.

Look at the output:

<img width="2034" alt="image" src="https://user-images.githubusercontent.com/894884/203626988-b771192c-08eb-4e0a-91df-f3db41c697d5.png">

**Expected behaviour**
No stack overflows :)

**Comment by kubukoz** [comment]
I guess it'll be an upstream change in https://github.com/VirtusLab/using_directives .

**Comment by tgodzik** [comment]
Thanks for reporting! This issue is caused by https://github.com/VirtusLab/using_directives/blob/90ee2eea1167b5f2ed63aa8158d4281e1eb905d8/src/main/java/com/virtuslab/using_directives/custom/Scanner.java#L456

@KacperFKorban Any chance you can take a look at it?

**Comment by KacperFKorban** [comment]
Sure, I'll take a look at it

**Comment by KacperFKorban** [comment]
Short update:
This specific case should be fixed by updating to `scala-cli` `0.1.18` (it uses the latest `using_directives` version)
Though in the latest version of `using_directives` there are still `StackOverflowErrors` caused by the same type of input (but require ~10 times bigger files i.e. 2600 loc). I'm working on the fix for that ATM.
